File: Sample/Engine/Curve.py
# -*- encoding: utf-8 -*-

from __future__ import annotations

import copy
from typing import Any, Dict, List


class Curve:
    r"""
    \brief Runtime curve asset backed by keyed interpolation data.

    Curve keys use UE-style time/value samples with interpolation mode and optional arrive/leave tangents.
    """

    def __init__(self, data: Dict[str, Any]) -> None:
        r"""
        \brief Construct a curve from asset data.

        - \param data Curve asset dictionary loaded from Data/Curves.
        """
        self.name: str = str(data.get("name", ""))
        self.defaultValue: float = float(data.get("defaultValue", 0.0) or 0.0)
        self.preInfinity: str = self._normaliseInfinityMode(data.get("preInfinity", "constant"))
        self.postInfinity: str = self._normaliseInfinityMode(data.get("postInfinity", "constant"))
        self.keys: List[Dict[str, Any]] = self._normaliseKeys(data.get("keys", []))

    @classmethod
    def fromData(cls, data: Dict[str, Any]) -> "Curve":
        r"""
        \brief Create a curve from asset data.

        - \param data Curve asset dictionary.
        - \return Curve instance.
        """
        return cls(data)

    def toData(self) -> Dict[str, Any]:
        r"""
        \brief Serialise this curve back to asset data.

        - \return Curve asset dictionary.
        """
        return {
            "type": "curve",
            "name": self.name,
            "defaultValue": self.defaultValue,
            "preInfinity": self.preInfinity,
            "postInfinity": self.postInfinity,
            "keys": copy.deepcopy(self.keys),
        }

    def isEmpty(self) -> bool:
        r"""
        \brief Check whether this curve has no keys.

        - \return True when the curve has no keyframes.
        """
        return len(self.keys) == 0

    def getDuration(self) -> float:
        r"""
        \brief Get the time span covered by the first and last key.

        - \return Curve duration in seconds or units.
        """
        if len(self.keys) < 2:
            return 0.0
        return float(self.keys[-1]["time"]) - float(self.keys[0]["time"])

    def evaluate(self, time: float) -> float:
        r"""
        \brief Evaluate the curve at the given time.

        - \param time Time coordinate.
        - \return Interpolated curve value.
        """
        if not self.keys:
            return self.defaultValue
        if len(self.keys) == 1:
            return float(self.keys[0]["value"])
        sampleTime = float(time)
        first = self.keys[0]
        last = self.keys[-1]
        if sampleTime <= float(first["time"]):
            return self._extrapolate(sampleTime, first, self.keys[1], self.preInfinity, True)
        if sampleTime >= float(last["time"]):
            return self._extrapolate(sampleTime, self.keys[-2], last, self.postInfinity, False)
        for index in range(len(self.keys) - 1):
            start = self.keys[index]
            end = self.keys[index + 1]
            startTime = float(start["time"])
            endTime = float(end["time"])
            if startTime <= sampleTime <= endTime:
                return self._evaluateSegment(start, end, sampleTime)
        return float(last["value"])

    def _normaliseInfinityMode(self, value: Any) -> str:
        mode = str(value)
        return mode if mode in ("constant", "linear") else "constant"

    def _normaliseInterpolation(self, value: Any) -> str:
        mode = str(value)
        return mode if mode in ("constant", "linear", "cubic") else "linear"

    def _normaliseKeys(self, value: Any) -> List[Dict[str, Any]]:
        if not isinstance(value, list):
            return []
        keys: List[Dict[str, Any]] = []
        for item in value:
            if not isinstance(item, dict):
                continue
            keys.append(
                {
                    "time": float(item.get("time", 0.0) or 0.0),
                    "value": float(item.get("value", 0.0) or 0.0),
                    "interpolation": self._normaliseInterpolation(item.get("interpolation", "linear")),
                    "arriveTangent": float(item.get("arriveTangent", 0.0) or 0.0),
                    "leaveTangent": float(item.get("leaveTangent", 0.0) or 0.0),
                }
            )
        return sorted(keys, key=lambda key: float(key["time"]))

    def _evaluateSegment(self, start: Dict[str, Any], end: Dict[str, Any], time: float) -> float:
        startTime = float(start["time"])
        endTime = float(end["time"])
        startValue = float(start["value"])
        endValue = float(end["value"])
        duration = endTime - startTime
        if duration <= 0.0:
            return endValue
        interpolation = str(start.get("interpolation", "linear"))
        if interpolation == "constant":
            return startValue
        alpha = (time - startTime) / duration
        if interpolation == "cubic":
            return self._cubicHermite(
                startValue,
                endValue,
                float(start.get("leaveTangent", 0.0) or 0.0) * duration,
                float(end.get("arriveTangent", 0.0) or 0.0) * duration,
                alpha,
            )
        return startValue + (endValue - startValue) * alpha

    def _cubicHermite(
        self,
        startValue: float,
        endValue: float,
        leaveTangent: float,
        arriveTangent: float,
        alpha: float,
    ) -> float:
        t2 = alpha * alpha
        t3 = t2 * alpha
        return (
            (2.0 * t3 - 3.0 * t2 + 1.0) * startValue
            + (t3 - 2.0 * t2 + alpha) * leaveTangent
            + (-2.0 * t3 + 3.0 * t2) * endValue
            + (t3 - t2) * arriveTangent
        )

    def _extrapolate(
        self, time: float, start: Dict[str, Any], end: Dict[str, Any], mode: str, beforeFirst: bool
    ) -> float:
        edgeKey = start if beforeFirst else end
        if mode != "linear":
            return float(edgeKey["value"])
        startTime = float(start["time"])
        endTime = float(end["time"])
        startValue = float(start["value"])
        endValue = float(end["value"])
        duration = endTime - startTime
        if duration <= 0.0:
            return float(edgeKey["value"])
        slope = (endValue - startValue) / duration
        return float(edgeKey["value"]) + (time - float(edgeKey["time"])) * slope

File: Sample/Engine/Source/Core/src/Curve.cpp
#include <Curve.hpp>
#include <CurveKey.hpp>

#include <algorithm>
#include <cmath>

Curve::Curve(Curve::CurveData data)
    : name(std::move(data.name)),
      defaultValue(data.defaultValue),
      preInfinity(normaliseInfinityMode(data.preInfinity)),
      postInfinity(normaliseInfinityMode(data.postInfinity)),
      keys(std::move(data.keys)) {
    for (CurveKey& key : keys) {
        key.interpolation = normaliseInterpolation(key.interpolation);
    }
    std::stable_sort(keys.begin(), keys.end(),
                     [](const CurveKey& left, const CurveKey& right) {
                         return left.time < right.time;
                     });
}

std::shared_ptr<Curve> Curve::fromData(const Curve::CurveData& data) {
    return std::make_shared<Curve>(data);
}

Curve::CurveData Curve::toData() const {
    return {"curve", name, defaultValue, preInfinity, postInfinity, keys};
}

bool Curve::isEmpty() const {
    return keys.empty();
}

float Curve::getDuration() const {
    return keys.size() < 2 ? 0.0f : keys.back().time - keys.front().time;
}

float Curve::evaluate(float time) const {
    if (keys.empty()) {
        return defaultValue;
    }
    if (keys.size() == 1) {
        return keys.front().value;
    }
    if (time <= keys.front().time) {
        return extrapolate(time, keys[0], keys[1], preInfinity, true);
    }
    if (time >= keys.back().time) {
        return extrapolate(time, keys[keys.size() - 2], keys.back(),
                           postInfinity, false);
    }
    const auto ending =
        std::lower_bound(keys.begin(), keys.end(), time,
                         [](const CurveKey& key, float sampleTime) {
                             return key.time < sampleTime;
                         });
    return evaluateSegment(*(ending - 1), *ending, time);
}

std::string Curve::normaliseInfinityMode(const std::string& mode) {
    return mode == "linear" ? "linear" : "constant";
}

std::string Curve::normaliseInterpolation(const std::string& mode) {
    if (mode == "constant" || mode == "cubic") {
        return mode;
    }
    return "linear";
}

float Curve::evaluateSegment(const CurveKey& start, const CurveKey& ending,
                             float time) {
    const float duration = ending.time - start.time;
    if (duration <= 0.0f) {
        return ending.value;
    }
    if (start.interpolation == "constant") {
        return start.value;
    }
    const float alpha = (time - start.time) / duration;
    if (start.interpolation != "cubic") {
        return start.value + (ending.value - start.value) * alpha;
    }
    const float alpha2 = alpha * alpha;
    const float alpha3 = alpha2 * alpha;
    const float leaveTangent = start.leaveTangent * duration;
    const float arriveTangent = ending.arriveTangent * duration;
    return (2.0f * alpha3 - 3.0f * alpha2 + 1.0f) * start.value +
           (alpha3 - 2.0f * alpha2 + alpha) * leaveTangent +
           (-2.0f * alpha3 + 3.0f * alpha2) * ending.value +
           (alpha3 - alpha2) * arriveTangent;
}

float Curve::extrapolate(float time, const CurveKey& start,
                         const CurveKey& ending, const std::string& mode,
                         bool beforeFirst) {
    const CurveKey& edge = beforeFirst ? start : ending;
    const float duration = ending.time - start.time;
    if (mode != "linear" || duration <= 0.0f) {
        return edge.value;
    }
    const float slope = (ending.value - start.value) / duration;
    return edge.value + (time - edge.time) * slope;
}
